🐛 BUG: Routes do not work if api token is restricted to zone

### What version of `Wrangler` are you using?

0.0.0-21ee93e

### What operating system are you using?

Linux

### Describe the Bug

1. Restrict API token to a zone, for example:
```
Account - Workers R2 Storage:Edit, Workers Tail:Read, Workers KV Storage:Edit, Workers Scripts:Edit, Account Settings:Read
    example.com - Workers Routes:Edit
All users - User Details:Read
```

2. Try to deploy to `routes = ["https://example.com/*"]`
3. Error:
```
 ⛅️ wrangler 0.0.0-21ee93e 
---------------------------
Uploaded test-worker (1.71 sec)

✘ [ERROR] Received a bad response from the API

  Authentication error [code: 10000]
```

Not sure if this is related with the zone inference or a bug on Cloudflare's side

Has anyone confirmed the fix for this? I'm using the "Edit Workers" template for the API Token via Github Actions, but this is not working. 

I am able to locally run `npx wrangler publish` without issue, but when deploying via github actions, it fails with this error. 
```
webpack 5.72.1 compiled successfully in 1966 ms
✘ [ERROR] ✘ [ERROR] Received a bad response from the API  
    Authentication error [code: 10000]    
    If you think this is a bug, please open an issue at: https://github.com/cloudflare/wrangler2/issues/new
```


See below for my config:
wrangler.toml
```
name = "XXX"
main = "dist/worker.js"
account_id = "YYY"
# route =
workers_dev = true
compatibility_date = "2022-05-04"

[build]
command = "npm install && npm run build"
```

ci.yml
```
name: API CI
on: [push]
jobs:
  deploy:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - name: Publish to Cloudflare
        uses: cloudflare/wrangler-action@2.0.0
        with:
          apiToken: ${{ secrets.CF_API_TOKEN }}
```

--- a/.changeset/honest-cobras-relate.md
+++ b/.changeset/honest-cobras-relate.md
@@ -1,0 +1,14 @@
+---
+"wrangler": patch
+---
+
+fix: fallback on old zone-based API when account-based route API fails
+
+While we wait for changes to the CF API to support API tokens that do not have
+"All Zone" permissions, this change provides a workaround for most scenarios.
+
+If the bulk-route request fails with an authorization error, then we fallback
+to the Wrangler 1 approach, which sends individual route updates via a zone-based
+endpoint.
+
+Fixes #651

--- a/packages/wrangler/src/__tests__/dev.test.tsx
+++ b/packages/wrangler/src/__tests__/dev.test.tsx
@@ -152,10 +152,12 @@ describe("wrangler dev", () => {
       fs.writeFileSync("index.js", `export default {};`);
       mockGetZones("some-host.com", [{ id: "some-zone-id" }]);
       await runWrangler("dev --host some-host.com");
-      expect((Dev as jest.Mock).mock.calls[0][0].zone).toEqual({
-        host: "some-host.com",
-        id: "some-zone-id",
-      });
+      expect((Dev as jest.Mock).mock.calls[0][0]).toEqual(
+        expect.objectContaining({
+          host: "some-host.com",
+          zone: "some-zone-id",
+        })
+      );
     });
 
     it("should read wrangler.toml's dev.host", async () => {
@@ -168,9 +170,7 @@ describe("wrangler dev", () => {
       fs.writeFileSync("index.js", `export default {};`);
       mockGetZones("some-host.com", [{ id: "some-zone-id" }]);
       await runWrangler("dev");
-      expect((Dev as jest.Mock).mock.calls[0][0].zone.host).toEqual(
-        "some-host.com"
-      );
+      expect((Dev as jest.Mock).mock.calls[0][0].host).toEqual("some-host.com");
     });
 
     it("should read --route", async () => {
@@ -180,9 +180,7 @@ describe("wrangler dev", () => {
       fs.writeFileSync("index.js", `export default {};`);
       mockGetZones("some-host.com", [{ id: "some-zone-id" }]);
       await runWrangler("dev --route http://some-host.com/some/path/*");
-      expect((Dev as jest.Mock).mock.calls[0][0].zone.host).toEqual(
-        "some-host.com"
-      );
+      expect((Dev as jest.Mock).mock.calls[0][0].host).toEqual("some-host.com");
     });
 
     it("should read wrangler.toml's routes", async () => {
@@ -196,9 +194,7 @@ describe("wrangler dev", () => {
       fs.writeFileSync("index.js", `export default {};`);
       mockGetZones("some-host.com", [{ id: "some-zone-id" }]);
       await runWrangler("dev");
-      expect((Dev as jest.Mock).mock.calls[0][0].zone.host).toEqual(
-        "some-host.com"
-      );
+      expect((Dev as jest.Mock).mock.calls[0][0].host).toEqual("some-host.com");
     });
 
     it("should read wrangler.toml's environment specific routes", async () => {
@@ -220,9 +216,7 @@ describe("wrangler dev", () => {
       fs.writeFileSync("index.js", `export default {};`);
       mockGetZones("some-host.com", [{ id: "some-zone-id" }]);
       await runWrangler("dev --env staging");
-      expect((Dev as jest.Mock).mock.calls[0][0].zone.host).toEqual(
-        "some-host.com"
-      );
+      expect((Dev as jest.Mock).mock.calls[0][0].host).toEqual("some-host.com");
     });
 
     it("should strip leading `*` from given host when deducing a zone id", async () => {
@@ -233,9 +227,7 @@ describe("wrangler dev", () => {
       fs.writeFileSync("index.js", `export default {};`);
       mockGetZones("some-host.com", [{ id: "some-zone-id" }]);
       await runWrangler("dev");
-      expect((Dev as jest.Mock).mock.calls[0][0].zone.host).toEqual(
-        "some-host.com"
-      );
+      expect((Dev as jest.Mock).mock.calls[0][0].host).toEqual("some-host.com");
     });
 
     it("should strip leading `*.` from given host when deducing a zone id", async () => {
@@ -246,9 +238,7 @@ describe("wrangler dev", () => {
       fs.writeFileSync("index.js", `export default {};`);
       mockGetZones("some-host.com", [{ id: "some-zone-id" }]);
       await runWrangler("dev");
-      expect((Dev as jest.Mock).mock.calls[0][0].zone.host).toEqual(
-        "some-host.com"
-      );
+      expect((Dev as jest.Mock).mock.calls[0][0].host).toEqual("some-host.com");
     });
 
     it("should, when provided, use a configured zone_id", async () => {
@@ -260,10 +250,12 @@ describe("wrangler dev", () => {
       });
       fs.writeFileSync("index.js", `export default {};`);
       await runWrangler("dev");
-      expect((Dev as jest.Mock).mock.calls[0][0].zone).toEqual({
-        host: "some-domain.com",
-        id: "some-zone-id",
-      });
+      expect((Dev as jest.Mock).mock.calls[0][0]).toEqual(
+        expect.objectContaining({
+          host: "some-domain.com",
+          zone: "some-zone-id",
+        })
+      );
     });
 
     it("should, when provided, use a zone_name to get a zone_id", async () => {
@@ -276,11 +268,13 @@ describe("wrangler dev", () => {
       fs.writeFileSync("index.js", `export default {};`);
       mockGetZones("some-zone.com", [{ id: "a-zone-id" }]);
       await runWrangler("dev");
-      expect((Dev as jest.Mock).mock.calls[0][0].zone).toEqual({
-        // note that it uses the provided zone_name as a host too
-        host: "some-zone.com",
-        id: "a-zone-id",
-      });
+      expect((Dev as jest.Mock).mock.calls[0][0]).toEqual(
+        expect.objectContaining({
+          // note that it uses the provided zone_name as a host too
+          host: "some-zone.com",
+          zone: "a-zone-id",
+        })
+      );
     });
 
     it("given a long host, it should use the longest subdomain that resolves to a zone", async () => {
@@ -292,10 +286,12 @@ describe("wrangler dev", () => {
       mockGetZones("222.333.some-host.com", []);
       mockGetZones("333.some-host.com", [{ id: "some-zone-id" }]);
       await runWrangler("dev --host 111.222.333.some-host.com");
-      expect((Dev as jest.Mock).mock.calls[0][0].zone).toEqual({
-        host: "111.222.333.some-host.com",
-        id: "some-zone-id",
-      });
+      expect((Dev as jest.Mock).mock.calls[0][0]).toEqual(
+        expect.objectContaining({
+          host: "111.222.333.some-host.com",
+          zone: "some-zone-id",
+        })
+      );
     });
 
     it("should, in order, use args.host/config.dev.host/args.routes/(config.route|config.routes)", async () => {
@@ -310,10 +306,12 @@ describe("wrangler dev", () => {
         routes: ["http://5.some-host.com/some/path/*"],
       });
       await runWrangler("dev");
-      expect((Dev as jest.Mock).mock.calls[0][0].zone).toEqual({
-        host: "5.some-host.com",
-        id: "some-zone-id-5",
-      });
+      expect((Dev as jest.Mock).mock.calls[0][0]).toEqual(
+        expect.objectContaining({
+          host: "5.some-host.com",
+          zone: "some-zone-id-5",
+        })
+      );
       (Dev as jest.Mock).mockClear();
 
       // config.route
@@ -323,10 +321,12 @@ describe("wrangler dev", () => {
         route: "https://4.some-host.com/some/path/*",
       });
       await runWrangler("dev");
-      expect((Dev as jest.Mock).mock.calls[0][0].zone).toEqual({
-        host: "4.some-host.com",
-        id: "some-zone-id-4",
-      });
+      expect((Dev as jest.Mock).mock.calls[0][0]).toEqual(
+        expect.objectContaining({
+          host: "4.some-host.com",
+          zone: "some-zone-id-4",
+        })
+      );
       (Dev as jest.Mock).mockClear();
 
       // --routes
@@ -336,10 +336,12 @@ describe("wrangler dev", () => {
         route: "https://4.some-host.com/some/path/*",
       });
       await runWrangler("dev --routes http://3.some-host.com/some/path/*");
-      expect((Dev as jest.Mock).mock.calls[0][0].zone).toEqual({
-        host: "3.some-host.com",
-        id: "some-zone-id-3",
-      });
+      expect((Dev as jest.Mock).mock.calls[0][0]).toEqual(
+        expect.objectContaining({
+          host: "3.some-host.com",
+          zone: "some-zone-id-3",
+        })
+      );
       (Dev as jest.Mock).mockClear();
 
       // config.dev.host
@@ -352,10 +354,12 @@ describe("wrangler dev", () => {
         route: "4.some-host.com/some/path/*",
       });
       await runWrangler("dev --routes http://3.some-host.com/some/path/*");
-      expect((Dev as jest.Mock).mock.calls[0][0].zone).toEqual({
-        host: "2.some-host.com",
-        id: "some-zone-id-2",
-      });
+      expect((Dev as jest.Mock).mock.calls[0][0]).toEqual(
+        expect.objectContaining({
+          host: "2.some-host.com",
+          zone: "some-zone-id-2",
+        })
+      );
       (Dev as jest.Mock).mockClear();
 
       // --host
@@ -370,10 +374,12 @@ describe("wrangler dev", () => {
       await runWrangler(
         "dev --routes http://3.some-host.com/some/path/* --host 1.some-host.com"
       );
-      expect((Dev as jest.Mock).mock.calls[0][0].zone).toEqual({
-        host: "1.some-host.com",
-        id: "some-zone-id-1",
-      });
+      expect((Dev as jest.Mock).mock.calls[0][0]).toEqual(
+        expect.objectContaining({
+          host: "1.some-host.com",
+          zone: "some-zone-id-1",
+        })
+      );
       (Dev as jest.Mock).mockClear();
     });
 

--- a/packages/wrangler/src/__tests__/publish.test.ts
+++ b/packages/wrangler/src/__tests__/publish.test.ts
@@ -649,8 +649,85 @@ describe("publish", () => {
       await runWrangler("publish ./index --env dev --legacy-env false");
     });
 
+    it("should fallback to the Wrangler 1 zone-based API if the bulk-routes API fails", async () => {
+      writeWranglerToml({
+        routes: ["example.com/some-route/*"],
+      });
+      writeWorkerSource();
+      mockUpdateWorkerRequest({ enabled: false });
+      mockUploadWorkerRequest({ expectedType: "esm" });
+      // Simulate the bulk-routes API failing with a not authorized error.
+      mockUnauthorizedPublishRoutesRequest();
+      // Simulate that the worker has already been deployed to another route in this zone.
+      mockCollectKnownRoutesRequest([
+        {
+          pattern: "foo.example.com/other-route",
+          script: "test-name",
+        },
+      ]);
+      mockGetZoneFromHostRequest("example.com", "some-zone-id");
+      mockPublishRoutesFallbackRequest({
+        pattern: "example.com/some-route/*",
+        script: "test-name",
+      });
+      await runWrangler("publish ./index");
+
+      expect(std.err).toMatchInlineSnapshot(`""`);
+      expect(std.warn).toMatchInlineSnapshot(`
+        "[33m▲ [43;33m[[43;30mWARNING[43;33m][0m [1mThe current authentication token does not have 'All Zones' permissions so cannot use the bulk-routes API endpoint.[0m
+
+          Falling back to using the zone-based API endpoint to update each route individually.
+          Note that there is no access to read or create routes associated with zones that the API token
+          does not have permission for.
+
+
+        [33m▲ [43;33m[[43;30mWARNING[43;33m][0m [1mPreviously deployed routes:[0m
+
+          The following routes were already associated with this worker, and have not been deleted:
+           - \\"foo.example.com/other-route\\"
+          If these routes are not wanted then you can remove them in the dashboard.
+          These routes would have been deleted automatically if the bulk-routes API had been used.
+
+        "
+      `);
+      expect(std.out).toMatchInlineSnapshot(`
+        "Uploaded test-name (TIMINGS)
+        Published test-name (TIMINGS)
+          example.com/some-route/*"
+      `);
+    });
+
+    it("should error if the bulk-routes API fails and trying to push to a non-production environment", async () => {
+      writeWranglerToml({
+        routes: ["example.com/some-route/*"],
+        legacy_env: false,
+      });
+      writeWorkerSource();
+      mockUpdateWorkerRequest({ env: "staging", enabled: false });
+      mockUploadWorkerRequest({ env: "staging", expectedType: "esm" });
+      // Simulate the bulk-routes API failing with a not authorized error.
+      mockUnauthorizedPublishRoutesRequest({ env: "staging" });
+      // Simulate that the worker has already been deployed to another route in this zone.
+      mockCollectKnownRoutesRequest([
+        {
+          pattern: "foo.example.com/other-route",
+          script: "test-name",
+        },
+      ]);
+      mockGetZoneFromHostRequest("example.com", "some-zone-id");
+      mockPublishRoutesFallbackRequest({
+        pattern: "example.com/some-route/*",
+        script: "test-name",
+      });
+      await expect(runWrangler("publish ./index --env=staging")).rejects
+        .toThrowErrorMatchingInlineSnapshot(`
+              "Service environments combined with an API token that doesn't have 'All Zones' permissions is not supported.
+              Either turn off service environments by setting \`legacy_env = true\`, creating an API token with 'All Zones' permissions, or logging in via OAuth"
+            `);
+    });
+
     describe("custom domains", () => {
-      it("should publish routes marked with 'custom_domain' as seperate custom domains", async () => {
+      it("should publish routes marked with 'custom_domain' as separate custom domains", async () => {
         writeWranglerToml({
           routes: [{ pattern: "api.example.com", custom_domain: true }],
         });
@@ -5004,7 +5081,7 @@ addEventListener('fetch', event => {});`
           "debug": "",
           "err": "",
           "out": "--dry-run: exiting now.",
-          "warn": "[33m▲ [43;33m[[43;30mWARNING[43;33m][0m [1mEnabling node.js compatibility mode for builtins and globals. This is experimental and has serious tradeoffs. Please see https://github.com/ionic-team/rollup-plugin-node-polyfills/ for more details.[0m
+          "warn": "[33m▲ [43;33m[[43;30mWARNING[43;33m][0m [1mEnabling node.js compatibility mode for built-ins and globals. This is experimental and has serious tradeoffs. Please see https://github.com/ionic-team/rollup-plugin-node-polyfills/ for more details.[0m
 
         ",
         }
@@ -5048,7 +5125,7 @@ addEventListener('fetch', event => {});`
           "debug": "",
           "err": "",
           "out": "--dry-run: exiting now.",
-          "warn": "[33m▲ [43;33m[[43;30mWARNING[43;33m][0m [1mEnabling node.js compatibility mode for builtins and globals. This is experimental and has serious tradeoffs. Please see https://github.com/ionic-team/rollup-plugin-node-polyfills/ for more details.[0m
+          "warn": "[33m▲ [43;33m[[43;30mWARNING[43;33m][0m [1mEnabling node.js compatibility mode for built-ins and globals. This is experimental and has serious tradeoffs. Please see https://github.com/ionic-team/rollup-plugin-node-polyfills/ for more details.[0m
 
         ",
         }
@@ -5224,6 +5301,49 @@ function mockPublishRoutesRequest({
       return null;
     }
   );
+}
+
+function mockUnauthorizedPublishRoutesRequest({
+  env = undefined,
+  legacyEnv = false,
+}: {
+  env?: string | undefined;
+  legacyEnv?: boolean | undefined;
+} = {}) {
+  const servicesOrScripts = env && !legacyEnv ? "services" : "scripts";
+  const environment = env && !legacyEnv ? "/environments/:envName" : "";
+
+  setMockRawResponse(
+    `/accounts/:accountId/workers/${servicesOrScripts}/:scriptName${environment}/routes`,
+    "PUT",
+    () =>
+      createFetchResult(null, false, [
+        { message: "Authentication error", code: 10000 },
+      ])
+  );
+}
+
+function mockCollectKnownRoutesRequest(
+  routes: { pattern: string; script: string }[]
+) {
+  setMockResponse(`/zones/:zoneId/workers/routes`, "GET", () => routes);
+}
+
+function mockGetZoneFromHostRequest(host: string, zone: string) {
+  setMockResponse("/zones", (_uri, _init, queryParams) => {
+    expect(queryParams.get("name")).toEqual(host);
+    return [{ id: zone }];
+  });
+}
+
+function mockPublishRoutesFallbackRequest(route: {
+  pattern: string;
+  script: string;
+}) {
+  setMockResponse(`/zones/:zoneId/workers/routes`, "POST", (_url, { body }) => {
+    expect(JSON.parse(body as string)).toEqual(route);
+    return route.pattern;
+  });
 }
 
 function mockPublishCustomDomainsRequest({

--- a/packages/wrangler/src/__tests__/publish.test.ts
+++ b/packages/wrangler/src/__tests__/publish.test.ts
@@ -674,11 +674,12 @@ describe("publish", () => {
 
       expect(std.err).toMatchInlineSnapshot(`""`);
       expect(std.warn).toMatchInlineSnapshot(`
-        "[33m▲ [43;33m[[43;30mWARNING[43;33m][0m [1mThe current authentication token does not have 'All Zones' permissions so cannot use the bulk-routes API endpoint.[0m
+        "[33m▲ [43;33m[[43;30mWARNING[43;33m][0m [1mThe current authentication token does not have 'All Zones' permissions.[0m
 
           Falling back to using the zone-based API endpoint to update each route individually.
           Note that there is no access to read or create routes associated with zones that the API token
           does not have permission for.
+          No existing routes will been deleted in this case.
 
 
         [33m▲ [43;33m[[43;30mWARNING[43;33m][0m [1mPreviously deployed routes:[0m
@@ -686,7 +687,6 @@ describe("publish", () => {
           The following routes were already associated with this worker, and have not been deleted:
            - \\"foo.example.com/other-route\\"
           If these routes are not wanted then you can remove them in the dashboard.
-          These routes would have been deleted automatically if the bulk-routes API had been used.
 
         "
       `);

--- a/packages/wrangler/src/cfetch/internal.ts
+++ b/packages/wrangler/src/cfetch/internal.ts
@@ -1,3 +1,4 @@
+import assert from "node:assert";
 import { fetch, Headers } from "undici";
 import { version as wranglerVersion } from "../../package.json";
 import { getEnvironmentVariableFactory } from "../environment-variables";
@@ -30,6 +31,10 @@ export async function fetchInternal<ResponseType>(
   queryParams?: URLSearchParams,
   abortSignal?: AbortSignal
 ): Promise<ResponseType> {
+  assert(
+    resource.startsWith("/"),
+    `CF API fetch - resource path must start with a "/" but got "${resource}"`
+  );
   await requireLoggedIn();
   const apiToken = requireApiToken();
   const headers = cloneHeaders(init.headers);

--- a/packages/wrangler/src/config/config.ts
+++ b/packages/wrangler/src/config/config.ts
@@ -184,7 +184,7 @@ export interface DevConfig {
   /**
    * Host to forward requests to, defaults to the host of the first route of project
    */
-  host: Config["route"];
+  host: string | undefined;
 }
 
 export type RawDevConfig = Partial<DevConfig>;

--- a/packages/wrangler/src/create-worker-preview.ts
+++ b/packages/wrangler/src/create-worker-preview.ts
@@ -140,7 +140,8 @@ async function createPreviewToken(
   return {
     value: preview_token,
     host:
-      ctx.host ?? worker.name
+      ctx.host ??
+      (worker.name
         ? `${
             worker.name
             // TODO: this should also probably have the env prefix
@@ -150,7 +151,7 @@ async function createPreviewToken(
             //   ? `${ctx.env}.${worker.name}`
             //   : worker.name
           }.${host.split(".").slice(1).join(".")}`
-        : host,
+        : host),
 
     inspectorUrl,
     prewarmUrl,

--- a/packages/wrangler/src/create-worker-preview.ts
+++ b/packages/wrangler/src/create-worker-preview.ts
@@ -63,7 +63,7 @@ async function sessionToken(
 ): Promise<CfPreviewToken> {
   const { accountId } = account;
   const initUrl = ctx.zone
-    ? `/zones/${ctx.zone.id}/workers/edge-preview`
+    ? `/zones/${ctx.zone}/workers/edge-preview`
     : `/accounts/${accountId}/workers/subdomain/edge-preview`;
 
   const { exchange_url } = await fetchResult<{ exchange_url: string }>(
@@ -139,19 +139,18 @@ async function createPreviewToken(
 
   return {
     value: preview_token,
-    host: ctx.zone
-      ? ctx.zone.host
-      : worker.name
-      ? `${
-          worker.name
-          // TODO: this should also probably have the env prefix
-          // but it doesn't appear to work yet, instead giving us the
-          // "There is nothing here yet" screen
-          // ctx.env && !ctx.legacyEnv
-          //   ? `${ctx.env}.${worker.name}`
-          //   : worker.name
-        }.${host.split(".").slice(1).join(".")}`
-      : host,
+    host:
+      ctx.host ?? worker.name
+        ? `${
+            worker.name
+            // TODO: this should also probably have the env prefix
+            // but it doesn't appear to work yet, instead giving us the
+            // "There is nothing here yet" screen
+            // ctx.env && !ctx.legacyEnv
+            //   ? `${ctx.env}.${worker.name}`
+            //   : worker.name
+          }.${host.split(".").slice(1).join(".")}`
+        : host,
 
     inspectorUrl,
     prewarmUrl,

--- a/packages/wrangler/src/dev/dev.tsx
+++ b/packages/wrangler/src/dev/dev.tsx
@@ -54,12 +54,8 @@ export type DevProps = {
   };
   env: string | undefined;
   legacyEnv: boolean;
-  zone:
-    | {
-        id: string;
-        host: string;
-      }
-    | undefined;
+  zone: string | undefined;
+  host: string | undefined;
 };
 
 export function DevImplementation(props: DevProps): JSX.Element {
@@ -197,6 +193,7 @@ function DevSession(props: DevSessionProps) {
       env={props.env}
       legacyEnv={props.legacyEnv}
       zone={props.zone}
+      host={props.host}
     />
   );
 }

--- a/packages/wrangler/src/dev/remote.tsx
+++ b/packages/wrangler/src/dev/remote.tsx
@@ -30,7 +30,8 @@ export function Remote(props: {
   usageModel: "bundled" | "unbound" | undefined;
   env: string | undefined;
   legacyEnv: boolean | undefined;
-  zone: { id: string; host: string } | undefined;
+  zone: string | undefined;
+  host: string | undefined;
 }) {
   assert(props.accountId, "accountId is required");
   assert(props.apiToken, "apiToken is required");
@@ -50,6 +51,7 @@ export function Remote(props: {
     env: props.env,
     legacyEnv: props.legacyEnv,
     zone: props.zone,
+    host: props.host,
   });
 
   usePreviewServer({
@@ -83,7 +85,8 @@ export function useWorker(props: {
   usageModel: "bundled" | "unbound" | undefined;
   env: string | undefined;
   legacyEnv: boolean | undefined;
-  zone: { id: string; host: string } | undefined;
+  zone: string | undefined;
+  host: string | undefined;
 }): CfPreviewToken | undefined {
   const {
     name,
@@ -176,7 +179,12 @@ export function useWorker(props: {
             accountId,
             apiToken,
           },
-          { env: props.env, legacyEnv: props.legacyEnv, zone: props.zone },
+          {
+            env: props.env,
+            legacyEnv: props.legacyEnv,
+            zone: props.zone,
+            host: props.host,
+          },
           abortController.signal
         )
       );
@@ -222,6 +230,7 @@ export function useWorker(props: {
     props.env,
     props.legacyEnv,
     props.zone,
+    props.host,
   ]);
   return token;
 }

--- a/packages/wrangler/src/index.tsx
+++ b/packages/wrangler/src/index.tsx
@@ -66,7 +66,7 @@ import {
   requireAuth,
 } from "./user";
 import { whoami } from "./whoami";
-import { getZoneFromHost, getZonesAndHostsForRoutes } from "./zones";
+import { getZoneIdFromHost, getZoneForRoute } from "./zones";
 
 import type { Config } from "./config";
 import type { TailCLIFilters } from "./tail";
@@ -1086,17 +1086,20 @@ function createCLIParser(argv: string[]) {
 
       // Compute zone info from the `host` and `route` args and config;
       let host = args.host || config.dev.host;
-      let zone: string | undefined;
+      let zoneId: string | undefined;
 
       if (!args.local) {
         if (host) {
-          zone = await getZoneFromHost(host);
+          zoneId = await getZoneIdFromHost(host);
         }
         const routes = args.routes || config.route || config.routes;
-        if (!zone && routes) {
-          ({ zone, host } = (
-            await getZonesAndHostsForRoutes(routes).next()
-          ).value);
+        if (!zoneId && routes) {
+          const firstRoute = Array.isArray(routes) ? routes[0] : routes;
+          const zone = await getZoneForRoute(firstRoute);
+          if (zone) {
+            zoneId = zone.id;
+            host = zone.host;
+          }
         }
       }
 
@@ -1176,7 +1179,7 @@ function createCLIParser(argv: string[]) {
           name={getScriptName(args, config)}
           entry={entry}
           env={args.env}
-          zone={zone}
+          zone={zoneId}
           host={host}
           rules={getRules(config)}
           legacyEnv={isLegacyEnv(config)}

--- a/packages/wrangler/src/index.tsx
+++ b/packages/wrangler/src/index.tsx
@@ -66,6 +66,7 @@ import {
   requireAuth,
 } from "./user";
 import { whoami } from "./whoami";
+import { getZoneFromHost, getZonesAndHostsForRoutes } from "./zones";
 
 import type { Config } from "./config";
 import type { TailCLIFilters } from "./tail";
@@ -1083,88 +1084,20 @@ function createCLIParser(argv: string[]) {
 
       // TODO: if worker_dev = false and no routes, then error (only for dev)
 
-      /**
-       * Given something that resembles a URL,
-       * try to extract a host from it
-       */
-      function getHost(urlLike: string): string | undefined {
-        // strip leading * / *.
-        urlLike = urlLike.replace(/^\*(\.)?/g, "");
-
-        if (
-          !(urlLike.startsWith("http://") || urlLike.startsWith("https://"))
-        ) {
-          urlLike = "http://" + urlLike;
-        }
-        return new URL(urlLike).host;
-      }
-
-      /**
-       * Given something that resembles a host,
-       * try to infer a zone id from it
-       */
-      async function getZoneId(host: string): Promise<string | undefined> {
-        const zones = await fetchResult<{ id: string }[]>(
-          `/zones`,
-          {},
-          new URLSearchParams({ name: host })
-        );
-        return zones[0]?.id;
-      }
-
-      // When we're given a host (in one of the above ways), we do 2 things:
-      // - We try to extract a host from it
-      // - We try to get a zone id from the host
-      //
-      // So it turns out it's particularly hard to get a 'valid' domain
-      // from a string, so we don't even try to validate TLDs, etc.
-      // Once we get something that looks like w.x.y.z-ish, we then try to
-      // get a zone id for it, by lopping off subdomains until we get a hit
-      // from the API. That's it!
-
-      let zone: { host: string; id: string } | undefined;
+      // Compute zone info from the `host` and `route` args and config;
+      let host = args.host || config.dev.host;
+      let zone: string | undefined;
 
       if (!args.local) {
-        const hostLike =
-          args.host ||
-          config.dev.host ||
-          (args.routes && args.routes[0]) ||
-          config.route ||
-          (config.routes && config.routes[0]);
-
-        let zoneId: string | undefined =
-          typeof hostLike === "object" && "zone_id" in hostLike
-            ? hostLike.zone_id
-            : undefined;
-
-        const host =
-          typeof hostLike === "string"
-            ? getHost(hostLike)
-            : typeof hostLike === "object"
-            ? "zone_name" in hostLike
-              ? getHost(hostLike.zone_name)
-              : getHost(hostLike.pattern)
-            : undefined;
-
-        const hostPieces =
-          typeof host === "string" ? host.split(".") : undefined;
-
-        while (!zoneId && hostPieces && hostPieces.length > 1) {
-          zoneId = await getZoneId(hostPieces.join("."));
-          hostPieces.shift();
+        if (host) {
+          zone = await getZoneFromHost(host);
         }
-
-        if (host && !zoneId) {
-          throw new Error(`Could not find zone for ${hostLike}`);
+        const routes = args.routes || config.route || config.routes;
+        if (!zone && routes) {
+          ({ zone, host } = (
+            await getZonesAndHostsForRoutes(routes).next()
+          ).value);
         }
-
-        zone =
-          typeof zoneId === "string" && typeof host === "string"
-            ? {
-                host,
-                id: zoneId,
-              }
-            : undefined;
       }
 
       const nodeCompat = args.nodeCompat ?? config.node_compat;
@@ -1244,6 +1177,7 @@ function createCLIParser(argv: string[]) {
           entry={entry}
           env={args.env}
           zone={zone}
+          host={host}
           rules={getRules(config)}
           legacyEnv={isLegacyEnv(config)}
           minify={args.minify ?? config.minify}
@@ -1672,6 +1606,7 @@ function createCLIParser(argv: string[]) {
           rules={getRules(config)}
           env={args.env}
           zone={undefined}
+          host={undefined}
           legacyEnv={isLegacyEnv(config)}
           build={config.build || {}}
           minify={undefined}

--- a/packages/wrangler/src/publish.ts
+++ b/packages/wrangler/src/publish.ts
@@ -4,13 +4,15 @@ import path from "node:path";
 import { URLSearchParams } from "node:url";
 import tmp from "tmp-promise";
 import { bundleWorker } from "./bundle";
-import { fetchResult } from "./cfetch";
+import { fetchListResult, fetchResult } from "./cfetch";
 import { printBindings } from "./config";
 import { createWorkerUploadForm } from "./create-worker-upload-form";
 import { confirm } from "./dialogs";
 import { getMigrationsToUpload } from "./durable";
 import { logger } from "./logger";
+import { ParseError } from "./parse";
 import { syncAssets } from "./sites";
+import { getZonesAndHostsForRoutes } from "./zones";
 import type { Config } from "./config";
 import type {
   Route,
@@ -21,6 +23,7 @@ import type {
 import type { Entry } from "./entry";
 import type { AssetPaths } from "./sites";
 import type { CfWorkerInit } from "./worker";
+import type { RouteZoneAndHost } from "./zones";
 
 type Props = {
   config: Config;
@@ -260,7 +263,7 @@ export default async function publish(props: Props): Promise<void> {
   const nodeCompat = props.nodeCompat ?? config.node_compat;
   if (nodeCompat) {
     logger.warn(
-      "Enabling node.js compatibility mode for builtins and globals. This is experimental and has serious tradeoffs. Please see https://github.com/ionic-team/rollup-plugin-node-polyfills/ for more details."
+      "Enabling node.js compatibility mode for built-ins and globals. This is experimental and has serious tradeoffs. Please see https://github.com/ionic-team/rollup-plugin-node-polyfills/ for more details."
     );
   }
 
@@ -291,7 +294,7 @@ export default async function publish(props: Props): Promise<void> {
   const envName = props.env ?? "production";
 
   const start = Date.now();
-  const notProd = !props.legacyEnv && props.env;
+  const notProd = Boolean(!props.legacyEnv && props.env);
   const workerName = notProd ? `${scriptName} (${envName})` : scriptName;
   const workerUrl = notProd
     ? `/accounts/${accountId}/workers/services/${scriptName}/environments/${envName}`
@@ -497,18 +500,7 @@ export default async function publish(props: Props): Promise<void> {
   // Update routing table for the script.
   if (routesOnly.length > 0) {
     deployments.push(
-      fetchResult(`${workerUrl}/routes`, {
-        // Note: PUT will delete previous routes on this script.
-        method: "PUT",
-        body: JSON.stringify(
-          routesOnly.map((route) =>
-            typeof route !== "object" ? { pattern: route } : route
-          )
-        ),
-        headers: {
-          "Content-Type": "application/json",
-        },
-      }).then(() => {
+      publishRoutes(workerUrl, routesOnly, scriptName, notProd).then(() => {
         if (routesOnly.length > 10) {
           return routesOnly
             .slice(0, 9)
@@ -580,4 +572,160 @@ async function getSubdomain(accountId: string): Promise<string> {
       throw e;
     }
   }
+}
+
+/**
+ * Associate the newly deployed Worker with the given routes.
+ */
+async function publishRoutes(
+  workerUrl: string,
+  routes: Route[],
+  scriptName: string,
+  notProd: boolean
+): Promise<string[]> {
+  try {
+    return await fetchResult(`${workerUrl}/routes`, {
+      // Note: PUT will delete previous routes on this script.
+      method: "PUT",
+      body: JSON.stringify(
+        routes.map((route) =>
+          typeof route !== "object" ? { pattern: route } : route
+        )
+      ),
+      headers: {
+        "Content-Type": "application/json",
+      },
+    });
+  } catch (e) {
+    if (isAuthenticationError(e)) {
+      // An authentication error is probably due to a known issue,
+      // where the user is logged in via an API token that does not have "All Zones".
+      return await publishRoutesFallback(routes, scriptName, notProd);
+    } else {
+      throw e;
+    }
+  }
+}
+
+/**
+ * Try updating routes for the Worker using a less optimal zone-based API.
+ *
+ * Compute match zones to the routes, then for each route attempt to connect it to the Worker via the zone.
+ */
+async function publishRoutesFallback(
+  routes: Route[],
+  scriptName: string,
+  notProd: boolean
+) {
+  if (notProd) {
+    throw new Error(
+      "Service environments combined with an API token that doesn't have 'All Zones' permissions is not supported.\n" +
+        "Either turn off service environments by setting `legacy_env = true`, creating an API token with 'All Zones' permissions, or logging in via OAuth"
+    );
+  }
+  logger.warn(
+    "The current authentication token does not have 'All Zones' permissions so cannot use the bulk-routes API endpoint.\n" +
+      "Falling back to using the zone-based API endpoint to update each route individually.\n" +
+      "Note that there is no access to read or create routes associated with zones that the API token does not have permission for."
+  );
+
+  const deployedRoutes: string[] = [];
+
+  // Collect the routes (and their zones) that will be deployed.
+  const activeZones = new Map<string, string>();
+  const routesToDeploy: RouteZoneAndHost[] = [];
+  for await (const routeToDeploy of getZonesAndHostsForRoutes(routes)) {
+    activeZones.set(routeToDeploy.zone, routeToDeploy.host);
+    routesToDeploy.push(routeToDeploy);
+  }
+
+  // Collect the routes that are already deployed.
+  const allRoutes = new Map<string, string>();
+  const alreadyDeployedRoutes = new Set<string>();
+  for (const [zone, host] of activeZones) {
+    await collectKnownRoutes(
+      zone,
+      host,
+      scriptName,
+      allRoutes,
+      alreadyDeployedRoutes
+    );
+  }
+
+  // Deploy each route that is not already deployed.
+  for (const { routePattern, zone } of routesToDeploy) {
+    if (allRoutes.has(routePattern)) {
+      const knownScript = allRoutes.get(routePattern);
+      if (knownScript === scriptName) {
+        // This route is already associated with this worker, so no need to hit the API.
+        alreadyDeployedRoutes.delete(routePattern);
+        continue;
+      } else {
+        throw new Error(
+          `The route with pattern "${routePattern}" is already associated with another worker called "${knownScript}".`
+        );
+      }
+    }
+
+    const { pattern } = await fetchResult(`/zones/${zone}/workers/routes`, {
+      method: "POST",
+      body: JSON.stringify({
+        pattern: routePattern,
+        script: scriptName,
+      }),
+      headers: {
+        "Content-Type": "application/json",
+      },
+    });
+
+    deployedRoutes.push(pattern);
+  }
+
+  if (alreadyDeployedRoutes.size) {
+    logger.warn(
+      "Previously deployed routes:\n" +
+        "The following routes were already associated with this worker, and have not been deleted:\n" +
+        [...alreadyDeployedRoutes.values()].map((route) => ` - "${route}"\n`) +
+        "If these routes are not wanted then you can remove them in the dashboard.\n" +
+        "These routes would have been deleted automatically if the bulk-routes API had been used."
+    );
+  }
+
+  return deployedRoutes;
+}
+
+/**
+ * For the given `zone`, add to the mappings of routes to script names, and the set of currently deployed routes.
+ */
+async function collectKnownRoutes(
+  zone: string,
+  host: string,
+  scriptName: string,
+  /** A map of routes to scripts. */
+  allRoutes: Map<string, string>,
+  /** A set of routes that are already deployed to the script being published. */
+  alreadyDeployedRoutes: Set<string>
+): Promise<void> {
+  try {
+    for (const { pattern, script } of await fetchListResult<{
+      pattern: string;
+      script: string;
+    }>(`/zones/${zone}/workers/routes`)) {
+      allRoutes.set(pattern, script);
+      if (script === scriptName) {
+        alreadyDeployedRoutes.add(pattern);
+      }
+    }
+  } catch (e) {
+    if (isAuthenticationError(e)) {
+      e.notes.push({
+        text: `This could be because the API token being used does not have permission to access the zone "${host}" (${zone}).`,
+      });
+    }
+    throw e;
+  }
+}
+
+function isAuthenticationError(e: unknown): e is ParseError {
+  return e instanceof ParseError && (e as { code?: number }).code === 10000;
 }

--- a/packages/wrangler/src/publish.ts
+++ b/packages/wrangler/src/publish.ts
@@ -12,7 +12,7 @@ import { getMigrationsToUpload } from "./durable";
 import { logger } from "./logger";
 import { ParseError } from "./parse";
 import { syncAssets } from "./sites";
-import { getZonesAndHostsForRoutes } from "./zones";
+import { getZoneForRoute } from "./zones";
 import type { Config } from "./config";
 import type {
   Route,
@@ -23,7 +23,6 @@ import type {
 import type { Entry } from "./entry";
 import type { AssetPaths } from "./sites";
 import type { CfWorkerInit } from "./worker";
-import type { RouteZoneAndHost } from "./zones";
 
 type Props = {
   config: Config;
@@ -633,10 +632,16 @@ async function publishRoutesFallback(
 
   // Collect the routes (and their zones) that will be deployed.
   const activeZones = new Map<string, string>();
-  const routesToDeploy: RouteZoneAndHost[] = [];
-  for await (const routeToDeploy of getZonesAndHostsForRoutes(routes)) {
-    activeZones.set(routeToDeploy.zone, routeToDeploy.host);
-    routesToDeploy.push(routeToDeploy);
+  const routesToDeploy = new Map<string, string>();
+  for (const route of routes) {
+    const zone = await getZoneForRoute(route);
+    if (zone) {
+      activeZones.set(zone.id, zone.host);
+      routesToDeploy.set(
+        typeof route === "string" ? route : route.pattern,
+        zone.id
+      );
+    }
   }
 
   // Collect the routes that are already deployed.
@@ -664,7 +669,7 @@ async function publishRoutesFallback(
   }
 
   // Deploy each route that is not already deployed.
-  for (const { routePattern, zone } of routesToDeploy) {
+  for (const [routePattern, zoneId] of routesToDeploy.entries()) {
     if (allRoutes.has(routePattern)) {
       const knownScript = allRoutes.get(routePattern);
       if (knownScript === scriptName) {
@@ -678,7 +683,7 @@ async function publishRoutesFallback(
       }
     }
 
-    const { pattern } = await fetchResult(`/zones/${zone}/workers/routes`, {
+    const { pattern } = await fetchResult(`/zones/${zoneId}/workers/routes`, {
       method: "POST",
       body: JSON.stringify({
         pattern: routePattern,

--- a/packages/wrangler/src/publish.ts
+++ b/packages/wrangler/src/publish.ts
@@ -623,9 +623,10 @@ async function publishRoutesFallback(
     );
   }
   logger.warn(
-    "The current authentication token does not have 'All Zones' permissions so cannot use the bulk-routes API endpoint.\n" +
+    "The current authentication token does not have 'All Zones' permissions.\n" +
       "Falling back to using the zone-based API endpoint to update each route individually.\n" +
-      "Note that there is no access to read or create routes associated with zones that the API token does not have permission for."
+      "Note that there is no access to read or create routes associated with zones that the API token does not have permission for.\n" +
+      "No existing routes will been deleted in this case."
   );
 
   const deployedRoutes: string[] = [];
@@ -702,8 +703,7 @@ async function publishRoutesFallback(
       "Previously deployed routes:\n" +
         "The following routes were already associated with this worker, and have not been deleted:\n" +
         [...alreadyDeployedRoutes.values()].map((route) => ` - "${route}"\n`) +
-        "If these routes are not wanted then you can remove them in the dashboard.\n" +
-        "These routes would have been deleted automatically if the bulk-routes API had been used."
+        "If these routes are not wanted then you can remove them in the dashboard."
     );
   }
 

--- a/packages/wrangler/src/publish.ts
+++ b/packages/wrangler/src/publish.ts
@@ -643,13 +643,24 @@ async function publishRoutesFallback(
   const allRoutes = new Map<string, string>();
   const alreadyDeployedRoutes = new Set<string>();
   for (const [zone, host] of activeZones) {
-    await collectKnownRoutes(
-      zone,
-      host,
-      scriptName,
-      allRoutes,
-      alreadyDeployedRoutes
-    );
+    try {
+      for (const { pattern, script } of await fetchListResult<{
+        pattern: string;
+        script: string;
+      }>(`/zones/${zone}/workers/routes`)) {
+        allRoutes.set(pattern, script);
+        if (script === scriptName) {
+          alreadyDeployedRoutes.add(pattern);
+        }
+      }
+    } catch (e) {
+      if (isAuthenticationError(e)) {
+        e.notes.push({
+          text: `This could be because the API token being used does not have permission to access the zone "${host}" (${zone}).`,
+        });
+      }
+      throw e;
+    }
   }
 
   // Deploy each route that is not already deployed.
@@ -692,38 +703,6 @@ async function publishRoutesFallback(
   }
 
   return deployedRoutes;
-}
-
-/**
- * For the given `zone`, add to the mappings of routes to script names, and the set of currently deployed routes.
- */
-async function collectKnownRoutes(
-  zone: string,
-  host: string,
-  scriptName: string,
-  /** A map of routes to scripts. */
-  allRoutes: Map<string, string>,
-  /** A set of routes that are already deployed to the script being published. */
-  alreadyDeployedRoutes: Set<string>
-): Promise<void> {
-  try {
-    for (const { pattern, script } of await fetchListResult<{
-      pattern: string;
-      script: string;
-    }>(`/zones/${zone}/workers/routes`)) {
-      allRoutes.set(pattern, script);
-      if (script === scriptName) {
-        alreadyDeployedRoutes.add(pattern);
-      }
-    }
-  } catch (e) {
-    if (isAuthenticationError(e)) {
-      e.notes.push({
-        text: `This could be because the API token being used does not have permission to access the zone "${host}" (${zone}).`,
-      });
-    }
-    throw e;
-  }
 }
 
 function isAuthenticationError(e: unknown): e is ParseError {

--- a/packages/wrangler/src/publish.ts
+++ b/packages/wrangler/src/publish.ts
@@ -499,7 +499,7 @@ export default async function publish(props: Props): Promise<void> {
   // Update routing table for the script.
   if (routesOnly.length > 0) {
     deployments.push(
-      publishRoutes(workerUrl, routesOnly, scriptName, notProd).then(() => {
+      publishRoutes(routesOnly, { workerUrl, scriptName, notProd }).then(() => {
         if (routesOnly.length > 10) {
           return routesOnly
             .slice(0, 9)
@@ -577,10 +577,12 @@ async function getSubdomain(accountId: string): Promise<string> {
  * Associate the newly deployed Worker with the given routes.
  */
 async function publishRoutes(
-  workerUrl: string,
   routes: Route[],
-  scriptName: string,
-  notProd: boolean
+  {
+    workerUrl,
+    scriptName,
+    notProd,
+  }: { workerUrl: string; scriptName: string; notProd: boolean }
 ): Promise<string[]> {
   try {
     return await fetchResult(`${workerUrl}/routes`, {
@@ -599,7 +601,7 @@ async function publishRoutes(
     if (isAuthenticationError(e)) {
       // An authentication error is probably due to a known issue,
       // where the user is logged in via an API token that does not have "All Zones".
-      return await publishRoutesFallback(routes, scriptName, notProd);
+      return await publishRoutesFallback(routes, { scriptName, notProd });
     } else {
       throw e;
     }
@@ -613,8 +615,7 @@ async function publishRoutes(
  */
 async function publishRoutesFallback(
   routes: Route[],
-  scriptName: string,
-  notProd: boolean
+  { scriptName, notProd }: { scriptName: string; notProd: boolean }
 ) {
   if (notProd) {
     throw new Error(

--- a/packages/wrangler/src/worker.ts
+++ b/packages/wrangler/src/worker.ts
@@ -176,5 +176,6 @@ export interface CfWorkerInit {
 export interface CfWorkerContext {
   env: string | undefined;
   legacyEnv: boolean | undefined;
-  zone: { id: string; host: string } | undefined;
+  zone: string | undefined;
+  host: string | undefined;
 }

--- a/packages/wrangler/src/zones.ts
+++ b/packages/wrangler/src/zones.ts
@@ -7,6 +7,7 @@ import type { Route } from "./config/environment";
 export interface RouteZoneAndHost {
   zone: string;
   host: string;
+  routePattern: string;
 }
 
 /**
@@ -46,6 +47,7 @@ export async function* getZonesAndHostsForRoutes(
       yield {
         host,
         zone,
+        routePattern: typeof route === "string" ? route : route.pattern,
       };
     }
   }

--- a/packages/wrangler/src/zones.ts
+++ b/packages/wrangler/src/zones.ts
@@ -1,0 +1,90 @@
+import { fetchListResult } from "./cfetch";
+import type { Route } from "./config/environment";
+
+/**
+ * An object holding information about a zone for publishing.
+ */
+export interface RouteZoneAndHost {
+  zone: string;
+  host: string;
+}
+
+/**
+ * Try to compute the a zone ID and host name for one or more routes.
+ *
+ * When we're given a route, we do 2 things:
+ * - We try to extract a host from it
+ * - We try to get a zone id from the host
+ */
+export async function* getZonesAndHostsForRoutes(
+  routes: Route | Route[]
+): AsyncGenerator<RouteZoneAndHost> {
+  if (!Array.isArray(routes)) {
+    routes = [routes];
+  }
+
+  for (const route of routes) {
+    let zone: string | undefined =
+      typeof route === "object" && "zone_id" in route
+        ? route.zone_id
+        : undefined;
+
+    const host =
+      typeof route === "string"
+        ? getHost(route)
+        : typeof route === "object"
+        ? "zone_name" in route
+          ? getHost(route.zone_name)
+          : getHost(route.pattern)
+        : undefined;
+
+    if (host && !zone) {
+      zone = await getZoneFromHost(host);
+    }
+
+    if (zone && host) {
+      yield {
+        host,
+        zone,
+      };
+    }
+  }
+}
+
+/**
+ * Given something that resembles a URL, try to extract a host from it.
+ */
+function getHost(urlLike: string): string | undefined {
+  // strip leading * / *.
+  urlLike = urlLike.replace(/^\*(\.)?/g, "");
+
+  if (!(urlLike.startsWith("http://") || urlLike.startsWith("https://"))) {
+    urlLike = "http://" + urlLike;
+  }
+  return new URL(urlLike).host;
+}
+
+/**
+ * Given something that resembles a host, try to infer a zone id from it.
+ *
+ * It's hard to get a 'valid' domain from a string, so we don't even try to validate TLDs, etc.
+ * For each domain-like part of the host (e.g. w.x.y.z) try to get a zone id for it by
+ * lopping off subdomains until we get a hit from the API.
+ */
+export async function getZoneFromHost(host: string): Promise<string> {
+  const hostPieces = host.split(".");
+
+  while (hostPieces.length > 1) {
+    const zones = await fetchListResult<{ id: string }>(
+      `/zones`,
+      {},
+      new URLSearchParams({ name: hostPieces.join(".") })
+    );
+    if (zones.length > 0) {
+      return zones[0].id;
+    }
+    hostPieces.shift();
+  }
+
+  throw new Error(`Could not find zone for ${host}`);
+}

--- a/packages/wrangler/src/zones.ts
+++ b/packages/wrangler/src/zones.ts
@@ -4,10 +4,9 @@ import type { Route } from "./config/environment";
 /**
  * An object holding information about a zone for publishing.
  */
-export interface RouteZoneAndHost {
-  zone: string;
+export interface Zone {
+  id: string;
   host: string;
-  routePattern: string;
 }
 
 /**
@@ -17,46 +16,28 @@ export interface RouteZoneAndHost {
  * - We try to extract a host from it
  * - We try to get a zone id from the host
  */
-export async function* getZonesAndHostsForRoutes(
-  routes: Route | Route[]
-): AsyncGenerator<RouteZoneAndHost> {
-  if (!Array.isArray(routes)) {
-    routes = [routes];
-  }
-
-  for (const route of routes) {
-    let zone: string | undefined =
-      typeof route === "object" && "zone_id" in route
-        ? route.zone_id
-        : undefined;
-
-    const host =
-      typeof route === "string"
-        ? getHost(route)
-        : typeof route === "object"
-        ? "zone_name" in route
-          ? getHost(route.zone_name)
-          : getHost(route.pattern)
-        : undefined;
-
-    if (host && !zone) {
-      zone = await getZoneFromHost(host);
-    }
-
-    if (zone && host) {
-      yield {
-        host,
-        zone,
-        routePattern: typeof route === "string" ? route : route.pattern,
-      };
-    }
-  }
+export async function getZoneForRoute(route: Route): Promise<Zone | undefined> {
+  const host =
+    typeof route === "string"
+      ? getHostFromUrl(route)
+      : typeof route === "object"
+      ? "zone_name" in route
+        ? getHostFromUrl(route.zone_name)
+        : getHostFromUrl(route.pattern)
+      : undefined;
+  const id =
+    typeof route === "object" && "zone_id" in route
+      ? route.zone_id
+      : host
+      ? await getZoneIdFromHost(host)
+      : undefined;
+  return id && host ? { id, host } : undefined;
 }
 
 /**
  * Given something that resembles a URL, try to extract a host from it.
  */
-function getHost(urlLike: string): string | undefined {
+function getHostFromUrl(urlLike: string): string | undefined {
   // strip leading * / *.
   urlLike = urlLike.replace(/^\*(\.)?/g, "");
 
@@ -73,7 +54,7 @@ function getHost(urlLike: string): string | undefined {
  * For each domain-like part of the host (e.g. w.x.y.z) try to get a zone id for it by
  * lopping off subdomains until we get a hit from the API.
  */
-export async function getZoneFromHost(host: string): Promise<string> {
+export async function getZoneIdFromHost(host: string): Promise<string> {
   const hostPieces = host.split(".");
 
   while (hostPieces.length > 1) {

